lib/model/rwfolder.go

```diff
@@ -1372,12 +1374,49 @@ func (f *rwFolder) dbUpdaterRoutine() {
 	tick := time.NewTicker(maxBatchTime)
 	defer tick.Stop()
 
+	var changedFiles []string
+	var changedDirs []string
+	if f.fsync {
+		changedFiles = make([]string, 0, maxBatchSize)
+		changedDirs = make([]string, 0, maxBatchSize)
+	}
+
+	syncFilesOnce := func(files []string, syncFn func(string) error) {
+		sort.Strings(files)
+		var lastFile string
+		for _, file := range files {
+			if lastFile != file {
+				lastFile = file
+				if err := syncFn(file); err != nil {
+					l.Infof("fsync %q failed: %v", file, err)
+				}
+			}
+		}
+	}
+
 	handleBatch := func() {
 		found := false
 		var lastFile protocol.FileInfo
 
 		for _, job := range batch {
 			files = append(files, job.file)
+			if f.fsync {
+				// collect changed files and dirs
+				switch job.jobType {
+				case dbUpdateShortcutFile:
```

Comment:
"shortcut file" means we just changed the metadata, i.e., timestamp or permissions, so there is no need to fsync the file. We might want to fsync the containing directory though if we're being pedanting about it, which I honestly don't think we need to be.